// @1inch/cross-chain-sdk/dist/esm/cross-chain-order/evm/evm-cross-chain-order.js
import { Interaction, MakerTraits, SettlementPostInteractionData, ZX } from '@1inch/fusion-sdk';
import assert from 'assert';
import { InnerOrder } from './inner-order.js';
import { EscrowExtension } from './escrow-extension.js';
import { AddressComplement } from '../../domains/addresses/address-complement.js';
import { now } from '../../utils/time/index.js';
import { createAddress, EvmAddress } from '../../domains/addresses/index.js';
import { BaseOrder } from '../base-order.js';
import { TRUE_ERC20 } from '../../deployments.js';
import { isEvm, isSupportedChain } from '../../chains.js';
import { bufferFromHex } from '../../utils/bytes.js';
export class EvmCrossChainOrder extends BaseOrder {
    inner;
    constructor(extension, orderInfo, extra) {
        super();
        this.inner = new InnerOrder(extension, orderInfo, extra);
    }
    get hashLock() {
        return this.escrowExtension.hashLockInfo;
    }
    get timeLocks() {
        return this.escrowExtension.timeLocks;
    }
    get srcSafetyDeposit() {
        return this.escrowExtension.srcSafetyDeposit;
    }
    get dstSafetyDeposit() {
        return this.escrowExtension.dstSafetyDeposit;
    }
    get dstChainId() {
        return this.inner.escrowExtension.dstChainId;
    }
    get escrowExtension() {
        return this.inner.escrowExtension;
    }
    get extension() {
        return this.inner.extension;
    }
    get maker() {
        return EvmAddress.fromString(this.inner.maker.toString());
    }
    get takerAsset() {
        return createAddress(this.inner.escrowExtension.dstToken.toString(), this.dstChainId);
    }
    get makerAsset() {
        return EvmAddress.fromString(this.inner.makerAsset.toString());
    }
    get takingAmount() {
        return this.inner.takingAmount;
    }
    get makingAmount() {
        return this.inner.makingAmount;
    }
    get salt() {
        return this.inner.salt;
    }
    /**
     * Real receiver of funds on dst chain
     */
    get receiver() {
        const receiver = createAddress(this.inner.receiver.toString(), this.dstChainId, this.escrowExtension.dstAddressFirstPart);
        return receiver.isZero() ? this.maker : receiver;
    }
    /**
     * Timestamp in sec
     */
    get deadline() {
        return this.inner.deadline;
    }
    /**
     * Timestamp in sec
     */
    get auctionStartTime() {
        return this.inner.auctionStartTime;
    }
    /**
     * Timestamp in sec
     */
    get auctionEndTime() {
        return this.inner.auctionEndTime;
    }
    get nonce() {
        return this.inner.nonce;
    }
    get partialFillAllowed() {
        return this.inner.partialFillAllowed;
    }
    get multipleFillsAllowed() {
        return this.inner.multipleFillsAllowed;
    }
    /**
     * Create new EvmCrossChainOrder
     */
    static new(escrowFactory, orderInfo, escrowParams, details, extra) {
        assert(isEvm(escrowParams.srcChainId), `Not supported chain ${escrowParams.srcChainId}`);
        assert(isSupportedChain(escrowParams.dstChainId), `Not supported chain ${escrowParams.dstChainId}`);
        assert(escrowParams.srcChainId !== escrowParams.dstChainId, 'Chains must be different');
        const postInteractionData = SettlementPostInteractionData.new({
            whitelist: details.whitelist.map((i) => ({
                address: i.address.inner,
                allowFrom: i.allowFrom
            })),
            resolvingStartTime: details.resolvingStartTime ?? BigInt(now())
        });
        if (!isEvm(escrowParams.dstChainId) && !orderInfo.receiver) {
            throw new Error('Receiver is required for non EVM chain');
        }
        const [complement, receiver] = orderInfo.receiver?.splitToParts() || [
            AddressComplement.ZERO,
            EvmAddress.ZERO
        ];
        const ext = new EscrowExtension(escrowFactory, details.auction, postInteractionData, extra?.permit
            ? new Interaction(orderInfo.makerAsset.inner, extra.permit)
            : undefined, escrowParams.hashLock, escrowParams.dstChainId, orderInfo.takerAsset, escrowParams.srcSafetyDeposit, escrowParams.dstSafetyDeposit, escrowParams.timeLocks, complement);
        return new EvmCrossChainOrder(ext, {
            ...orderInfo,
            receiver,
            takerAsset: TRUE_ERC20[escrowParams.srcChainId]
        }, extra);
    }
    /**
     * Create CrossChainOrder from order data and extension
     *
     */
    static fromDataAndExtension(order, extension) {
        const ext = EscrowExtension.fromExtension(extension);
        const makerTraits = new MakerTraits(BigInt(order.makerTraits));
        const deadline = makerTraits.expiration();
        const orderExpirationDelay = deadline === null
            ? undefined
            : deadline -
                ext.auctionDetails.startTime -
                ext.auctionDetails.duration;
        return new EvmCrossChainOrder(ext, {
            makerAsset: EvmAddress.fromUnknown(order.makerAsset),
            takerAsset: EvmAddress.fromUnknown(order.takerAsset),
            makingAmount: BigInt(order.makingAmount),
            takingAmount: BigInt(order.takingAmount),
            receiver: EvmAddress.fromUnknown(order.receiver),
            maker: EvmAddress.fromUnknown(order.maker),
            salt: BigInt(order.salt) >> 160n
        }, {
            enablePermit2: makerTraits.isPermit2(),
            nonce: makerTraits.nonceOrEpoch(),
            permit: extension.makerPermit === ZX
                ? undefined
                : Interaction.decode(extension.makerPermit).data,
            orderExpirationDelay,
            allowMultipleFills: makerTraits.isMultipleFillsAllowed(),
            allowPartialFills: makerTraits.isPartialFillAllowed()
        });
    }
    build() {
        return this.inner.build();
    }
    toJSON() {
        return this.build();
    }
    getOrderHash(srcChainId) {
        return this.inner.getOrderHash(srcChainId);
    }
    getOrderHashBuffer(srcChainId) {
        return bufferFromHex(this.getOrderHash(srcChainId));
    }
    getTypedData(srcChainId) {
        return this.inner.getTypedData(srcChainId);
    }
    getCalculator() {
        return this.inner.getCalculator();
    }
    /**
     * Check if `wallet` can fill order before other
     */
    isExclusiveResolver(wallet) {
        return this.inner.isExclusiveResolver(wallet.inner);
    }
    /**
     * Check if the auction has exclusive resolver, and it is in the exclusivity period
     *
     * @param time timestamp to check, `now()` by default
     */
    isExclusivityPeriod(time) {
        return this.inner.isExclusivityPeriod(time);
    }
    /**
     * Check whether address allowed to execute order at the given time
     *
     * @param executor address of executor
     * @param executionTime timestamp in sec at which order planning to execute
     */
    canExecuteAt(executor, executionTime) {
        return this.inner.canExecuteAt(executor.inner, executionTime);
    }
}
