// @1inch/cross-chain-sdk/dist/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.idls = exports.utils = exports.SVM_ESCROW_SRC_IDL = exports.SVM_ESCROW_DST_IDL = exports.NetworkEnum = exports.SupportedChains = exports.Web3ProviderConnector = exports.PrivateKeyProviderConnector = exports.randBigInt = exports.now = exports.calcMakingAmount = exports.calcTakingAmount = exports.AmountMode = exports.TakerTraits = exports.LimitOrderContract = exports.AuctionCalculator = exports.Interaction = exports.SettlementPostInteractionData = exports.Extension = exports.MakerTraits = exports.Address = void 0;
var fusion_sdk_1 = require("@1inch/fusion-sdk");
Object.defineProperty(exports, "Address", { enumerable: true, get: function () { return fusion_sdk_1.Address; } });
Object.defineProperty(exports, "MakerTraits", { enumerable: true, get: function () { return fusion_sdk_1.MakerTraits; } });
Object.defineProperty(exports, "Extension", { enumerable: true, get: function () { return fusion_sdk_1.Extension; } });
Object.defineProperty(exports, "SettlementPostInteractionData", { enumerable: true, get: function () { return fusion_sdk_1.SettlementPostInteractionData; } });
Object.defineProperty(exports, "Interaction", { enumerable: true, get: function () { return fusion_sdk_1.Interaction; } });
Object.defineProperty(exports, "AuctionCalculator", { enumerable: true, get: function () { return fusion_sdk_1.AuctionCalculator; } });
// Execution
Object.defineProperty(exports, "LimitOrderContract", { enumerable: true, get: function () { return fusion_sdk_1.LimitOrderContract; } });
Object.defineProperty(exports, "TakerTraits", { enumerable: true, get: function () { return fusion_sdk_1.TakerTraits; } });
Object.defineProperty(exports, "AmountMode", { enumerable: true, get: function () { return fusion_sdk_1.AmountMode; } });
// helpers
Object.defineProperty(exports, "calcTakingAmount", { enumerable: true, get: function () { return fusion_sdk_1.calcTakingAmount; } });
Object.defineProperty(exports, "calcMakingAmount", { enumerable: true, get: function () { return fusion_sdk_1.calcMakingAmount; } });
Object.defineProperty(exports, "now", { enumerable: true, get: function () { return fusion_sdk_1.now; } });
Object.defineProperty(exports, "randBigInt", { enumerable: true, get: function () { return fusion_sdk_1.randBigInt; } });
// connectors
Object.defineProperty(exports, "PrivateKeyProviderConnector", { enumerable: true, get: function () { return fusion_sdk_1.PrivateKeyProviderConnector; } });
Object.defineProperty(exports, "Web3ProviderConnector", { enumerable: true, get: function () { return fusion_sdk_1.Web3ProviderConnector; } });
var chains_js_1 = require("./chains.js");
Object.defineProperty(exports, "SupportedChains", { enumerable: true, get: function () { return chains_js_1.SupportedChains; } });
Object.defineProperty(exports, "NetworkEnum", { enumerable: true, get: function () { return chains_js_1.NetworkEnum; } });
__exportStar(require("./deployments.js"), exports);
__exportStar(require("./sdk/index.js"), exports);
__exportStar(require("./api/index.js"), exports);
__exportStar(require("./ws-api/index.js"), exports);
var cross_chain_escrow_dst_js_1 = require("./idl/cross-chain-escrow-dst.js");
Object.defineProperty(exports, "SVM_ESCROW_DST_IDL", { enumerable: true, get: function () { return cross_chain_escrow_dst_js_1.IDL; } });
var cross_chain_escrow_src_js_1 = require("./idl/cross-chain-escrow-src.js");
Object.defineProperty(exports, "SVM_ESCROW_SRC_IDL", { enumerable: true, get: function () { return cross_chain_escrow_src_js_1.IDL; } });
__exportStar(require("./domains/index.js"), exports);
__exportStar(require("./contracts/index.js"), exports);
__exportStar(require("./cross-chain-order/index.js"), exports);
exports.utils = __importStar(require("./utils/index.js"));
exports.idls = __importStar(require("./idl/index.js"));
