// @1inch/cross-chain-sdk/dist/cjs/api/orders/orders.api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrdersApi = void 0;
const orders_request_js_1 = require("./orders.request.js");
const params_js_1 = require("../params.js");
class OrdersApi {
    config;
    httpClient;
    static Version = 'v1.1';
    constructor(config, httpClient) {
        this.config = config;
        this.httpClient = httpClient;
    }
    async getActiveOrders(params = new orders_request_js_1.ActiveOrdersRequest()) {
        const queryParams = (0, params_js_1.concatQueryParams)(params.build());
        const url = `${this.config.url}/${OrdersApi.Version}/order/active/${queryParams}`;
        return this.httpClient.get(url);
    }
    async getOrderStatus(params) {
        const url = `${this.config.url}/${OrdersApi.Version}/order/status/${params.orderHash}`;
        return this.httpClient.get(url);
    }
    async getOrdersByMaker(params) {
        const qp = (0, params_js_1.concatQueryParams)(params.buildQueryParams());
        const url = `${this.config.url}/${OrdersApi.Version}/order/maker/${params.address}/${qp}`;
        return this.httpClient.get(url);
    }
    async getReadyToAcceptSecretFills(orderHash) {
        const url = `${this.config.url}/${OrdersApi.Version}/order/ready-to-accept-secret-fills/${orderHash}`;
        return this.httpClient.get(url);
    }
    async getReadyToExecutePublicActions() {
        const url = `${this.config.url}/${OrdersApi.Version}/order/ready-to-execute-public-actions`;
        return this.httpClient.get(url);
    }
    async getPublishedSecrets(orderHash) {
        const url = `${this.config.url}/${OrdersApi.Version}/order/secrets/${orderHash}`;
        return this.httpClient.get(url);
    }
    async getCancellableOrders(pagination) {
        const qp = {};
        if (pagination?.page !== undefined) {
            qp.page = pagination.page;
        }
        if (pagination?.limit) {
            qp.limit = pagination.limit;
        }
        const url = `${this.config.url}/${OrdersApi.Version}/order/cancelable-by-resolvers${(0, params_js_1.concatQueryParams)(qp)}`;
        return this.httpClient.get(url);
    }
}
exports.OrdersApi = OrdersApi;
