// @1inch/cross-chain-sdk/dist/esm/api/fusion-api.js
import { AxiosProviderConnector } from '@1inch/fusion-sdk';
import { QuoterApi } from './quoter/index.js';
import { RelayerApi } from './relayer/index.js';
import { ActiveOrdersRequest, OrdersApi } from './orders/index.js';
export class FusionApi {
    quoterApi;
    relayerApi;
    ordersApi;
    constructor(config) {
        const httpProvider = config.httpProvider || new AxiosProviderConnector(config.authKey);
        this.quoterApi = new QuoterApi({
            url: `${config.url}/quoter`,
            authKey: config.authKey
        }, httpProvider);
        this.relayerApi = new RelayerApi({
            url: `${config.url}/relayer`,
            authKey: config.authKey
        }, httpProvider);
        this.ordersApi = new OrdersApi({
            url: `${config.url}/orders`,
            authKey: config.authKey
        }, httpProvider);
    }
    getQuote(params) {
        return this.quoterApi.getQuote(params);
    }
    getQuoteWithCustomPreset(params, body) {
        return this.quoterApi.getQuoteWithCustomPreset(params, body);
    }
    getActiveOrders(params = new ActiveOrdersRequest()) {
        return this.ordersApi.getActiveOrders(params);
    }
    getOrderStatus(params) {
        return this.ordersApi.getOrderStatus(params);
    }
    getOrdersByMaker(params) {
        return this.ordersApi.getOrdersByMaker(params);
    }
    getReadyToAcceptSecretFills(orderHash) {
        return this.ordersApi.getReadyToAcceptSecretFills(orderHash);
    }
    getReadyToExecutePublicActions() {
        return this.ordersApi.getReadyToExecutePublicActions();
    }
    getPublishedSecrets(orderHash) {
        return this.ordersApi.getPublishedSecrets(orderHash);
    }
    getCancellableOrders(pagination) {
        return this.ordersApi.getCancellableOrders(pagination);
    }
    submitOrder(params) {
        return this.relayerApi.submit(params);
    }
    submitOrderBatch(params) {
        return this.relayerApi.submitBatch(params);
    }
    submitSecret(orderHash, secret) {
        return this.relayerApi.submitSecret(orderHash, secret);
    }
}
