// @1inch/cross-chain-sdk/dist/cjs/contracts/svm/svm-dst-escrow-factory.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SvmDstEscrowFactory = void 0;
const anchor_1 = require("@coral-xyz/anchor");
const assert_1 = __importDefault(require("assert"));
const instruction_js_1 = require("./instruction.js");
const base_program_js_1 = require("./base-program.js");
const whitelist_js_1 = require("./whitelist.js");
const bigint_to_bn_js_1 = require("../../utils/numbers/bigint-to-bn.js");
const uint_as_be_bytes_js_1 = require("../../utils/numbers/uint-as-be-bytes.js");
const uint256_split_js_1 = require("../../utils/numbers/uint256-split.js");
const index_js_1 = require("../../utils/index.js");
const index_js_2 = require("../../domains/index.js");
const cross_chain_escrow_dst_js_1 = require("../../idl/cross-chain-escrow-dst.js");
const bytes_js_1 = require("../../utils/bytes.js");
const bn_array_to_big_int_js_1 = require("../../utils/numbers/bn-array-to-big-int.js");
const bn_js_1 = require("../../utils/numbers/bn.js");
class SvmDstEscrowFactory extends base_program_js_1.BaseProgram {
    static DEFAULT = new SvmDstEscrowFactory(new index_js_2.SolanaAddress('AMEAktCrii7mVFQKCM9i5hKES4YrV3zFagrawr8BY8pb'));
    static coder = new anchor_1.BorshCoder(cross_chain_escrow_dst_js_1.IDL);
    constructor(programId) {
        super(programId);
    }
    static parseCreateEscrowInstruction(ix) {
        const decoded = SvmDstEscrowFactory.coder.instruction.decode(ix.data);
        (0, assert_1.default)(decoded, 'cannot decode create instruction');
        (0, assert_1.default)(decoded.name === 'create', 'not create instruction');
        const data = decoded.data;
        return {
            orderHash: (0, bytes_js_1.bufferToHex)(data.orderHash),
            hashlock: index_js_2.HashLock.fromString((0, bytes_js_1.bufferToHex)(data.hashlock)),
            amount: BigInt(data.amount.toString()),
            safetyDeposit: BigInt(data.safetyDeposit.toString()),
            recipient: index_js_2.SolanaAddress.fromBuffer(data.recipient.toBuffer()),
            timelocks: index_js_2.TimeLocks.fromBigInt((0, bn_array_to_big_int_js_1.bnArrayToBigInt)(data.timelocks)),
            srcCancellationTimestamp: data.srcCancellationTimestamp,
            assetIsNative: data.assetIsNative,
            taker: ix.accounts[0].pubkey,
            token: ix.accounts[1].pubkey,
            escrow: ix.accounts[3].pubkey
        };
    }
    static parsePrivateWithdrawInstruction(ix) {
        const decoded = this.coder.instruction.decode(ix.data);
        (0, assert_1.default)(decoded, 'cannot decode withdraw instruction');
        (0, assert_1.default)(decoded.name === 'withdraw', 'not withdraw instruction');
        return {
            secret: (0, bytes_js_1.bufferToHex)(decoded.data.secret)
        };
    }
    static parsePublicWithdrawInstruction(ix) {
        const decoded = this.coder.instruction.decode(ix.data);
        (0, assert_1.default)(decoded, 'cannot decode publicWithdraw instruction');
        (0, assert_1.default)(decoded.name === 'publicWithdraw', 'not publicWithdraw instruction');
        return {
            secret: (0, bytes_js_1.bufferToHex)(decoded.data.secret)
        };
    }
    getEscrowAddress(params) {
        return (0, index_js_1.getPda)(this.programId, [
            this.encoder.encode('escrow'),
            params.orderHash,
            params.hashLock.toBuffer(),
            params.taker.toBuffer(),
            (0, uint_as_be_bytes_js_1.uintAsBeBytes)(params.amount, 64)
        ]);
    }
    createEscrow(params, extra) {
        const token = params.token.isNative()
            ? index_js_2.SolanaAddress.WRAPPED_NATIVE
            : params.token;
        const data = SvmDstEscrowFactory.coder.instruction.encode('create', {
            orderHash: params.orderHash,
            hashlock: params.hashLock.toBuffer(),
            amount: new bn_js_1.BN(params.amount.toString()),
            safetyDeposit: new bn_js_1.BN(params.safetyDeposit.toString()),
            recipient: params.maker,
            timelocks: (0, uint256_split_js_1.uint256split)(params.timeLocks.build()).map(bigint_to_bn_js_1.bigintToBN),
            srcCancellationTimestamp: Number(extra.srcCancellationTimestamp),
            assetIsNative: params.token.isNative()
        });
        const escrowAddress = this.getEscrowAddress(params);
        return new instruction_js_1.Instruction(this.programId, [
            {
                // 1. taker
                pubkey: params.taker,
                isSigner: true,
                isWritable: true
            },
            {
                // 2. mint
                pubkey: token,
                isWritable: false,
                isSigner: false
            },
            this.optionalAccount({
                // 3. taker_ata
                pubkey: (0, index_js_1.getAta)(params.taker, token, extra.tokenProgramId),
                isSigner: false,
                isWritable: true
            }, params.token.isNative()),
            {
                // 4. escrow
                pubkey: escrowAddress,
                isSigner: false,
                isWritable: true
            },
            {
                // 5. escrow ata
                pubkey: (0, index_js_1.getAta)(escrowAddress, token, extra.tokenProgramId),
                isWritable: true,
                isSigner: false
            },
            {
                // 6. associated_token_program
                pubkey: index_js_2.SolanaAddress.ASSOCIATED_TOKE_PROGRAM_ID,
                isSigner: false,
                isWritable: false
            },
            {
                // 7. token_program
                pubkey: extra.tokenProgramId,
                isSigner: false,
                isWritable: false
            },
            {
                // 8. rent
                pubkey: index_js_2.SolanaAddress.SYSVAR_RENT_ID,
                isSigner: false,
                isWritable: false
            },
            {
                // 9. system_program
                pubkey: index_js_2.SolanaAddress.SYSTEM_PROGRAM_ID,
                isSigner: false,
                isWritable: false
            }
        ], data);
    }
    withdrawPrivate(params, secret, extra) {
        const token = params.token.isNative()
            ? index_js_2.SolanaAddress.WRAPPED_NATIVE
            : params.token;
        const data = SvmDstEscrowFactory.coder.instruction.encode('withdraw', {
            secret
        });
        const escrow = this.getEscrowAddress(params);
        return new instruction_js_1.Instruction(this.programId, [
            {
                // 1. taker
                pubkey: params.taker,
                isSigner: true,
                isWritable: true
            },
            {
                // 2. recipient
                pubkey: params.maker,
                isSigner: false,
                isWritable: true
            },
            {
                // 3. dst token
                pubkey: token,
                isSigner: false,
                isWritable: false
            },
            {
                // 4. escrow
                pubkey: escrow,
                isSigner: false,
                isWritable: true
            },
            {
                // 5. escrow_ata
                pubkey: (0, index_js_1.getAta)(escrow, token, extra.tokenProgramId),
                isSigner: false,
                isWritable: true
            },
            this.optionalAccount({
                // 6. recipient_ata
                pubkey: (0, index_js_1.getAta)(params.maker, params.token, extra.tokenProgramId),
                isSigner: false,
                isWritable: true
            }, params.token.isNative()),
            {
                // 7. associated_token_program
                pubkey: index_js_2.SolanaAddress.ASSOCIATED_TOKE_PROGRAM_ID,
                isSigner: false,
                isWritable: false
            },
            {
                // 8. token_program
                pubkey: extra.tokenProgramId,
                isSigner: false,
                isWritable: false
            },
            {
                // 9. system_program
                pubkey: index_js_2.SolanaAddress.SYSTEM_PROGRAM_ID,
                isSigner: false,
                isWritable: false
            }
        ], data);
    }
    withdrawPublic(params, secret, payer, extra) {
        const whitelistProgram = extra.whitelistProgramId
            ? new whitelist_js_1.WhitelistContract(extra.whitelistProgramId)
            : whitelist_js_1.WhitelistContract.DEFAULT;
        const token = params.token.isNative()
            ? index_js_2.SolanaAddress.WRAPPED_NATIVE
            : params.token;
        const data = SvmDstEscrowFactory.coder.instruction.encode('publicWithdraw', { secret });
        const escrow = this.getEscrowAddress(params);
        return new instruction_js_1.Instruction(this.programId, [
            {
                // 1. creator
                pubkey: params.taker,
                isSigner: false,
                isWritable: true
            },
            {
                // 2. recipient
                pubkey: params.maker,
                isSigner: false,
                isWritable: true
            },
            {
                // 3. payer
                pubkey: payer,
                isSigner: true,
                isWritable: true
            },
            {
                // 4. resolver_access
                pubkey: whitelistProgram.getAccessAccount(payer),
                isSigner: false,
                isWritable: false
            },
            {
                // 5. mint
                pubkey: token,
                isSigner: false,
                isWritable: false
            },
            {
                // 6. escrow
                pubkey: escrow,
                isSigner: false,
                isWritable: true
            },
            {
                // 7. escrow_ata
                pubkey: (0, index_js_1.getAta)(escrow, token, extra.tokenProgramId),
                isSigner: false,
                isWritable: true
            },
            this.optionalAccount({
                // 8. recipient_ata
                pubkey: (0, index_js_1.getAta)(params.maker, params.token, extra.tokenProgramId),
                isSigner: false,
                isWritable: true
            }, params.token.isNative()),
            {
                // 9. associated_token_program
                pubkey: index_js_2.SolanaAddress.ASSOCIATED_TOKE_PROGRAM_ID,
                isSigner: false,
                isWritable: false
            },
            {
                // 10. token_program
                pubkey: extra.tokenProgramId,
                isSigner: false,
                isWritable: false
            },
            {
                // 11. system_program
                pubkey: index_js_2.SolanaAddress.SYSTEM_PROGRAM_ID,
                isSigner: false,
                isWritable: false
            }
        ], data);
    }
    cancelPrivate(params, extra) {
        const token = params.token.isNative()
            ? index_js_2.SolanaAddress.WRAPPED_NATIVE
            : params.token;
        const data = SvmDstEscrowFactory.coder.instruction.encode('cancel', {});
        const escrow = this.getEscrowAddress(params);
        return new instruction_js_1.Instruction(this.programId, [
            {
                // 1. creator
                pubkey: params.taker,
                isSigner: true,
                isWritable: true
            },
            {
                // 2. mint
                pubkey: token,
                isSigner: false,
                isWritable: false
            },
            {
                // 3. escrow
                pubkey: escrow,
                isSigner: false,
                isWritable: true
            },
            {
                // 4. escrow_ata
                pubkey: (0, index_js_1.getAta)(escrow, token, extra.tokenProgramId),
                isSigner: false,
                isWritable: true
            },
            this.optionalAccount({
                // 5. creator_ata
                pubkey: (0, index_js_1.getAta)(params.taker, params.token, extra.tokenProgramId),
                isSigner: false,
                isWritable: true
            }, params.token.isNative()),
            {
                // 6. token_program
                pubkey: extra.tokenProgramId,
                isSigner: false,
                isWritable: false
            },
            {
                // 7. system_program
                pubkey: index_js_2.SolanaAddress.SYSTEM_PROGRAM_ID,
                isSigner: false,
                isWritable: false
            }
        ], data);
    }
}
exports.SvmDstEscrowFactory = SvmDstEscrowFactory;
