// @1inch/cross-chain-sdk/dist/cjs/api/fusion-api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FusionApi = void 0;
const fusion_sdk_1 = require("@1inch/fusion-sdk");
const index_js_1 = require("./quoter/index.js");
const index_js_2 = require("./relayer/index.js");
const index_js_3 = require("./orders/index.js");
class FusionApi {
    quoterApi;
    relayerApi;
    ordersApi;
    constructor(config) {
        const httpProvider = config.httpProvider || new fusion_sdk_1.AxiosProviderConnector(config.authKey);
        this.quoterApi = new index_js_1.QuoterApi({
            url: `${config.url}/quoter`,
            authKey: config.authKey
        }, httpProvider);
        this.relayerApi = new index_js_2.RelayerApi({
            url: `${config.url}/relayer`,
            authKey: config.authKey
        }, httpProvider);
        this.ordersApi = new index_js_3.OrdersApi({
            url: `${config.url}/orders`,
            authKey: config.authKey
        }, httpProvider);
    }
    getQuote(params) {
        return this.quoterApi.getQuote(params);
    }
    getQuoteWithCustomPreset(params, body) {
        return this.quoterApi.getQuoteWithCustomPreset(params, body);
    }
    getActiveOrders(params = new index_js_3.ActiveOrdersRequest()) {
        return this.ordersApi.getActiveOrders(params);
    }
    getOrderStatus(params) {
        return this.ordersApi.getOrderStatus(params);
    }
    getOrdersByMaker(params) {
        return this.ordersApi.getOrdersByMaker(params);
    }
    getReadyToAcceptSecretFills(orderHash) {
        return this.ordersApi.getReadyToAcceptSecretFills(orderHash);
    }
    getReadyToExecutePublicActions() {
        return this.ordersApi.getReadyToExecutePublicActions();
    }
    getPublishedSecrets(orderHash) {
        return this.ordersApi.getPublishedSecrets(orderHash);
    }
    getCancellableOrders(pagination) {
        return this.ordersApi.getCancellableOrders(pagination);
    }
    submitOrder(params) {
        return this.relayerApi.submit(params);
    }
    submitOrderBatch(params) {
        return this.relayerApi.submitBatch(params);
    }
    submitSecret(orderHash, secret) {
        return this.relayerApi.submitSecret(orderHash, secret);
    }
}
exports.FusionApi = FusionApi;
