// @1inch/cross-chain-sdk/dist/esm/api/orders/orders.request.js
import { PaginationRequest } from '../pagination.js';
export class ActiveOrdersRequest {
    pagination;
    constructor(params = {}) {
        this.pagination = new PaginationRequest(params.page, params.limit);
    }
    build() {
        return {
            page: this.pagination.page,
            limit: this.pagination.limit
        };
    }
}
export class OrderStatusRequest {
    orderHash;
    constructor(params) {
        this.orderHash = params.orderHash;
    }
    build() {
        return {
            orderHash: this.orderHash
        };
    }
}
export class OrdersByMakerRequest {
    address;
    pagination;
    srcChain;
    dstChain;
    srcToken;
    dstToken;
    withToken;
    timestampFrom;
    timestampTo;
    constructor(params) {
        this.address = params.address;
        this.pagination = new PaginationRequest(params.page, params.limit);
        this.srcChain = params.srcChain;
        this.dstChain = params.dstChain;
        this.srcToken = params.srcToken;
        this.dstToken = params.dstToken;
        this.withToken = params.withToken;
        this.timestampFrom = params.timestampFrom;
        this.timestampTo = params.timestampTo;
    }
    buildQueryParams() {
        return {
            limit: this.pagination.limit,
            page: this.pagination.page,
            srcChain: this.srcChain,
            dstChain: this.dstChain,
            srcToken: this.srcToken,
            dstToken: this.dstToken,
            withToken: this.withToken,
            timestampFrom: this.timestampFrom,
            timestampTo: this.timestampTo
        };
    }
}
