// @1inch/cross-chain-sdk/dist/esm/chains.js
export var NetworkEnum;
(function (NetworkEnum) {
    NetworkEnum[NetworkEnum["ETHEREUM"] = 1] = "ETHEREUM";
    NetworkEnum[NetworkEnum["POLYGON"] = 137] = "POLYGON";
    NetworkEnum[NetworkEnum["ZKSYNC"] = 324] = "ZKSYNC";
    NetworkEnum[NetworkEnum["BINANCE"] = 56] = "BINANCE";
    NetworkEnum[NetworkEnum["ARBITRUM"] = 42161] = "ARBITRUM";
    NetworkEnum[NetworkEnum["AVALANCHE"] = 43114] = "AVALANCHE";
    NetworkEnum[NetworkEnum["OPTIMISM"] = 10] = "OPTIMISM";
    NetworkEnum[NetworkEnum["FANTOM"] = 250] = "FANTOM";
    NetworkEnum[NetworkEnum["GNOSIS"] = 100] = "GNOSIS";
    NetworkEnum[NetworkEnum["COINBASE"] = 8453] = "COINBASE";
    NetworkEnum[NetworkEnum["LINEA"] = 59144] = "LINEA";
    NetworkEnum[NetworkEnum["SONIC"] = 146] = "SONIC";
    NetworkEnum[NetworkEnum["UNICHAIN"] = 130] = "UNICHAIN";
    NetworkEnum[NetworkEnum["SOLANA"] = 501] = "SOLANA";
})(NetworkEnum || (NetworkEnum = {}));
export const SupportedChains = [
    NetworkEnum.ETHEREUM,
    NetworkEnum.POLYGON,
    NetworkEnum.BINANCE,
    NetworkEnum.OPTIMISM,
    NetworkEnum.ARBITRUM,
    NetworkEnum.AVALANCHE,
    NetworkEnum.GNOSIS,
    NetworkEnum.COINBASE,
    NetworkEnum.ZKSYNC,
    NetworkEnum.LINEA,
    NetworkEnum.SONIC,
    NetworkEnum.UNICHAIN,
    NetworkEnum.SOLANA
];
export const isSupportedChain = (chain) => SupportedChains.includes(chain);
export const isEvm = (chain) => {
    return (SupportedChains.includes(chain) &&
        chain !== NetworkEnum.SOLANA);
};
export const isSolana = (chain) => {
    return chain === NetworkEnum.SOLANA;
};
